core/shared/src/main/scala/scalaz/zio/STM.scala

```diff
@@ -0,0 +1,226 @@
+
+import java.util.concurrent.atomic.AtomicLong
+
+/**
+ * `STM[E, A]` represents a computation that can be performed transactional
+ * resulting in a failure `E` or a value `A`.
+ *
+ * {{{
+ * def transfer(receiver: TVar[Int],
+ *              sender: TVar[Int], much: Int): IO[Nothing, Int] =
+ *   STM.atomically {
+ *     for {
+ *       _ <- receiver.update(_ + much)
+ *       s <- sender.get
+ *       _ <- check(s >= much)
+ *       _ <- sender.update(_ - much)
+ *       r <- receiver.get
+ *     } yield r
+ *   }
+ *
+ *   val action: IO[Nothing, Int] =
+ *     for {
+ *       t <- STM.atomically(TVar(0).seq(TVar(20000)))
+ *       (receiver, sender) = t
+ *       balance <- transfer(receiver, sender, 1000)
+ *     } yield balance
+ * }}}
+ */
+final case class STM[+E, +A] private (
+  run: Map[Long, STM.internal.Entry] => (Map[Long, STM.internal.Entry], Either[E, A])
+) { self =>
+
+  /**
+   * Maps from one error type to another.
+   */
+  def leftMap[E2](f: E => E2): STM[E2, A] =
+    STM(
+      journal =>
+        self.run(journal) match {
+          case (journal, either) => {
+            val leftMapped: Either[A, E2] = either.swap match {
+              case Right(e) => Right(f(e))
+              case Left(a)  => Left(a)
+            }
+            (journal, leftMapped.swap)
+          }
+      }
+    )
+
+  /**
+   * Maps the value produced by the computation.
+   */
+  def map[B](f: A => B): STM[E, B] =
+    STM(
+      journal =>
+        self.run(journal) match {
+          case (journal, either) => {
+            val eitherMapped = either match {
+              case Right(a) => Right(f(a))
+              case Left(b)  => Left(b)
+            }
+            (journal, eitherMapped)
+          }
+      }
+    )
+
+  /**
+   * Feeds the value produced by this computation to the specified computation,
+   * and then runs that computation as well to produce its failure or value.
+   */
+  def flatMap[E1 >: E, B](f: A => STM[E1, B]): STM[E1, B] =
+    STM(
+      journal =>
+        self.run(journal) match {
+          case (journal, Left(error)) => (journal, Left(error))
+          case (journal, Right(a))    => f(a).run(journal)
+      }
+    )
+
+  /**
+   * Tries this computation first, and if it fails, tries the other computation.
+   */
+  def orElse[E2, A1 >: A](that: => STM[E2, A1]): STM[E2, A1] =
+    self.attempt.flatMap(either => either.fold(_ => that, STM.point(_)))
+
+  /**
+   * Surfaces any errors at the value level, where they can be recovered from.
+   */
+  def attempt: STM[Nothing, Either[E, A]] =
+    STM(
+      journal =>
+        self.run(journal) match {
+          case (journal, value) => (journal, Right(value))
+      }
+    )
+
+  /**
+   * Sequentially zips this value with the specified one.
+   */
+  def seq[E1 >: E, B](that: => STM[E1, B]): STM[E1, (A, B)] =
+    self.flatMap(a => that.map(b => (a, b)))
+}
+
+object STM {
+
+  object internal {
+    def freshIdentity(): Long = counter.incrementAndGet()
```

Comment:
what's the worst thing that can happen if this wraps around? and how quickly can i reach this point on a modern machine?

Reply:
@ritschwumm I think I didn't get it. you meant how many times the function freshIdentity could possibly invoked?

Reply:
what i mean is: what happens (worst case) when i create a fresh identity that already has been created in the past? and how likely is it to occur under stress?

Reply:
Would we lose something if the journal keys were the TVar themselves, with reference equality and default hash code?

Reply:
We can switch to using something like `new java.lang.Object` and using Java identity map, which would guarantee no wraparound issues and probably be faster.

The wraparound for `Long` is not really an issue, though (it's more the performance). To wrap around in 1 year, you'd need to create 292,471,208,678 `TVar` per second.

Reply:
Hmm, that works. We just need to use Java's identity map.

@rajcspsg Let's use the TVar as the key using reference identity. So we just need Java's identity map.

Reply:
No, we won't need a counter at all.

We won't need fresh identity.

Rather, you can change `STM`:

```scala
final case class STM[+E, +A] private (
   run: IdentityHashMap[TVar[_], STM.internal.Entry] => (IdentityHashMap[TVar[_], STM.internal.Entry], Either[E, A])
 ) { self =>
```

Note that `Entry` can be modified to remove the `TVar` since it's already the key in the map. It only needs the new value and new version, I think.